DPP-1047_FixDataLockEventsRaceCondition

lazcool opened the pull request:
Change order of DataLockEvents table copying from transient to Deds, to fix issue that surfaces in the Commitments web job where we process DataLockEvents before their errors are associated with the events.
Bonus: fix build of solution

lazcool committed:
Change order of DataLockEvent & related tables copying from transient to deds
+ Fix build of ProviderPayments.TestStack solution
+ add ignored test useful for manually testing DataLockEvent table transfers

lazcool committed:
Merge branch 'develop' into DPP-1047_FixDataLockEventsRaceCondition

chrisfoster76 commented on src/AcceptanceTesting/Common/ProviderPayments.TestStack.Infrastructure/packages.config:
```diff
@@ -1,5 +1,6 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
+  <package id="AutoMapper" version="5.1.1" targetFramework="net45" />
```
Is this package required? Automapper is something we've generally avoided here.

lazcool replied:
It was added as the project doesn't build without it.
I've just checked and it appears to be actually used. Further investigation might find that it's not actually required, but I'd suggest that deeper investigation is outside of the scope of this fix.

chrisfoster76 commented on src/AcceptanceTesting/Common/ProviderPayments.TestStack.UI/ProviderPayments.TestStack.UI.csproj:
```diff
@@ -210,11 +210,6 @@
     <Compile Include="Controllers\PaymentReportController.cs" />
     <Compile Include="Controllers\ProcessController.cs" />
     <Compile Include="Controllers\ComponentAdminController.cs" />
-    <Compile Include="DependencyResolution\ControllerConvention.cs" />
```
Are these removals correct?

lazcool replied:
The files are not present in the repo at the location specified in the csprof file, which causes the build to fail. They appear to have been moved to a new location.

lazcool committed:
tidy comment only

lazcool committed:
Merge branch 'DPP-1047_FixDataLockEventsRaceCondition' of https://github.com/SkillsFundingAgency/das-providerpayments into DPP-1047_FixDataLockEventsRaceCondition

chrisfoster76 commented on src/SharedPipelineComponents/DataLockEvents/SFA.DAS.Provider.Events.DataLock.IntegrationTests/Transfer/WhenDataLockEventsAreTransferred.cs:
```diff
@@ -0,0 +1,38 @@
+{
+    class WhenDataLockEventsAreTransferred
+    {
+        //// comment out in GlobalSetup...
```
Is this an instruction to anyone who wants to run the test? In that case, could you just pop it into a comment block above the test, ie. "Remove BeforeAllTests() in GlobalSetup before running".

lazcool replied:
I've tidied up the file, including better instructions on how to run the test.

chrisfoster76 commented on src/SharedPipelineComponents/DataLockEvents/SFA.DAS.Provider.Events.DataLock.IntegrationTests/Transfer/WhenDataLockEventsAreTransferred.cs:
```diff
@@ -0,0 +1,38 @@
+        [Ignore("Useful for manual testing only (without further work)")]
+        public void ThenItShouldTransfer()
+        {
+            //TestDataHelper.SubmissionCopyReferenceData();
```
Could this commented line come out?

lazcool replied:
Gone!

chrisfoster76 commented on src/SharedPipelineComponents/DataLockEvents/SFA.DAS.Provider.Events.DataLock.IntegrationTests/Transfer/WhenDataLockEventsAreTransferred.cs:
```diff
@@ -0,0 +1,38 @@
+        {
+            //TestDataHelper.SubmissionCopyReferenceData();
+
+            const string mappingFilePath = @"C:\git\das-providerpayments\src\SharedPipelineComponents\DataLockEvents\DeployDataLock\copy mappings\DasDataLockEventsCopyToDedsMapping.xml";
```
I can see an issue with this path for purposes of running across different dev environments.

lazcool replied:
I've added a comment to that effect.

chrisfoster76 reviewed (commented):
Perhaps inclusion of the test is debatable (please see comments).

lazcool committed:
tidy up ignored WhenDataLockEventsAreTransferred unit test

lazcool commented:
I agree inclusion of the test is debatable, but if someone has to revisit the mapping file, they might find the test useful (I would!).

davehowlett reviewed (approved):
talked through change with Phil - looks fine

lazcool committed:
Fix DataLock tests (& actual mapping) specifically:
    * Add TransactionTypesFlag to DataLockEventPeriods in mapping file (went missing during merge)
Fix acceptance test startup generally
  * move index creation on Payments.Payments from PaymentsDue to LevyPayments
  * remove index creation on DataLock.[DataLockEventPeriods|DataLockEventCommitmentVersions] tables from DataLock (tables and indices are created in DataLockEvents)

patelmaulik commented on src/PeriodEnd/LevyPayments/DeployLevyPayments/sql/Migration Scripts/1_LevyPayments.deds.Add_Indexes.sql:
```diff
@@ -1,4 +1,10 @@
-
```
IF EXISTS - DROP INDEX  check required ?

lazcool replied:
Done, although as mentioned on Slack, we might want to change the check pattern we use for index creation to...
"if not exists, create"
but if all the indexes are on the transient db, rather than deds, it's not an issue

carlbm commented on src/AcceptanceTesting/Common/ProviderPayments.TestStack.UI/ProviderPayments.TestStack.UI.csproj:
```diff
@@ -319,8 +314,8 @@
     </PropertyGroup>
-    <Error Condition="!Exists('..\packages\Microsoft.Net.Compilers.1.0.0\build\Microsoft.Net.Compilers.props')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\Microsoft.Net.Compilers.1.0.0\build\Microsoft.Net.Compilers.props'))" />
-    <Error Condition="!Exists('..\packages\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.1.0.0\build\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.props')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.1.0.0\build\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.props'))" />
+    <Error Condition="!Exists('..\..\..\..\packages\Microsoft.Net.Compilers.1.0.0\build\Microsoft.Net.Compilers.props')" Text="$([System.String]::Format('$(ErrorText)', '..\..\..\..\packages\Microsoft.Net.Compilers.1.0.0\build\Microsoft.Net.Compilers.props'))" />
```
These two lines are repeated. Worth removing them

lazcool replied:
They were repeated twice, but have now removed the dupes.

carlbm reviewed (changes requested):
Can you add the check on the index before creating it

lazcool committed:
Add back index creation on DataLock.[DataLockEventPeriods|DataLockEventCommitmentVersions] under DataLockEvents (where the tables are created)
+ add 'if exists - drop' to index creation for Payments.Payments
+ remove deadwood from ProviderPayments.TestStack.UI.csproj

lazcool commented:
After double checking that the index creation on DataLock.[DataLockEventPeriods|DataLockEventCommitmentVersions] was still there (after on offline comment by Carl), I noticed that the change I thought I'd made had gone missing, so they have been added back with the latest commit.

lazcool commented:
After double checking that the index creation on DataLock.[DataLockEventPeriods|DataLockEventCommitmentVersions] was still there (after on offline comment by @carlbm), I noticed that the change I thought I'd made had gone missing, so they have been added back with the latest commit.

patelmaulik commented on src/SharedPipelineComponents/DataLockEvents/SFA.DAS.Provider.Events.DataLock.IntegrationTests/SFA.DAS.Provider.Events.DataLock.IntegrationTests.csproj:
```diff
@@ -145,6 +145,9 @@
     </Content>
+    <Content Include="..\DeployDataLock\sql\migrationscripts\005_DataLockEvents_Tables_Add_Indexes.sql">
+      <Link>DbSetupScripts\MigrationScripts\005_DataLockEvents_Tables_Add_Indexes.sql</Link>
+    </Content>
```
<CopyToOutputDirectory>Always</CopyToOutputDirectory>  ?

patelmaulik replied:
` <CopyToOutputDirectory>Always</CopyToOutputDirectory>`

patelmaulik commented on src/SharedPipelineComponents/DataLockEvents/SFA.DAS.Provider.Events.DataLock.IntegrationTests/SFA.DAS.Provider.Events.DataLock.IntegrationTests.csproj:
```diff
@@ -145,6 +145,9 @@
     </Content>
+    <Content Include="..\DeployDataLock\sql\migrationscripts\005_DataLockEvents_Tables_Add_Indexes.sql">
+      <Link>DbSetupScripts\MigrationScripts\005_DataLockEvents_Tables_Add_Indexes.sql</Link>
+    </Content>
```
` <CopyToOutputDirectory>Always</CopyToOutputDirectory>`

lazcool replied:
Change made and Data Lock and Data Lock Events integration tests successfully run.

patelmaulik reviewed (changes requested):
CopyAways needed for the new file.

lazcool committed:
new sql script needs to be copied to test dir

satya-sanikommu commented:
Testing completed and No regression issues. Ready to merge
